lib/gem_package_enumerator.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+##
+# A wrapper around Gem::Package to enumerate files using a lazy enumerator.
+#
+# Gem::Package#each_file iterates over a single IO stream.
+# In order to make the code more readable elsewhere and to ensure that we read
+# files one at a time, use a lazy enumerator wherever possible.
+
+class GemPackageEnumerator
```

Comment:
What about to keep the most interesting part (laziness) of this custom enumerator to be part of the name like `GemPackageLazyEnumerator`? IMHO the best would be to keep it as `Gem:Package::LazyEnumerator`. :thinking: 

Also is the comment above still proper? If I understand it well, there is no `Gem::Package#each_file` method anymore.

Reply:
`Gem::Package::LazyEnumerator` works, but ultimately I want to add `each_file` to `Gem::Package`, so this naming discussion might be moot. This class would disappear if I had each_file, but I didn't want to delay this PR while I worked on it.

Reply:
Merging this back upstream would be the best, but I was about to not block this until that happens and is released.

Reply:
I think it makes sense to me to keep the name as-is since this provides the base enumerator as well as (optional) laziness.

Reply:
@indirect Even the name would be the same, is the comment wrong?

Reply:
Updating the comment to be clear that it adds an enumerator, and provides a lazy enumerator if you do not provide a block, sounds very reasonable to me. 👍🏻

Reply:
Left the name, cleaned the comment up. Hopefully this goes away if I can upstream the each_entry method, or something like that.